clang/include/clang/CIR/Dialect/IR/CIROps.td

```diff
@@ -3854,6 +3854,42 @@ def MemCpyOp : CIR_Op<"libc.memcpy"> {
   }];
 }
 
+//===----------------------------------------------------------------------===//
+// MemMoveOp
+//===----------------------------------------------------------------------===//
+
+def MemMoveOp : CIR_Op<"libc.memmove", [AllTypesMatch<["dst", "src"]>]> {
```

Comment:
This is the same as `MemCpyOp` + overlapping semantics, I'm inclined to think that we should just have a `memory_overlap` UnitAttr in `MemCpyOp` instead. OTOH, it's nice that we have an explicit name to differentiate them (while matching the actual libc function name).

That said, I'm fine with this approach but we should unify the implementation bit on a new tablegen class, see `class CIR_CallOp` that is used for both `CallOp` and `TryCallOp` for an example. Since this will require updating the `MemCpyOp` with some constraints (e.g. `VoidPtr`, etc), it's fine if you do that in a follow up PR. I'm happy to merge this as soon as I have your confirmation on addressing this as a follow up!

Reply:
> unify the implementation bit on a new tablegen class, see class CIR_CallOp that is used for both CallOp and TryCallOp for an example

I tried to adopt this approach, as it results in a much cleaner implementation. However, I’m not quite sure how the `memory_overlap` UnitAttr is intended to function in this context. I’ve defined `MemCpyOp` and `MemMoveOp`, but I noticed that the `memory_overlap` UnitAttr is not being utilized.

Considering support for the [llvm.memcpy.inline.* intrinsics](https://llvm.org/docs/LangRef.html#llvm-memcpy-inline-intrinsic), we can reuse the `class CIR_MemCpyOp` and define `MemCpyInlineOp`. Should we also add an `inline` UnitAttr in this case? 

Additionally, I would like to ask if we should unify the `assemblyFormat`. Currently, `MemCpyOp` uses the following format:

```
`:` type($len) `` `,` qualified(type($src)) `->` qualified(type($dst))
```

Whereas `MemSetOp` employs:

```
`:` qualified(type($dst)) `,` type($len)
```

Would it be beneficial to standardize these formats?

Reply:
I prefer to use distinct operations for better clarity. We can support `__builtin_memcpy`, `__builtin_memmove`, and `__builtin_memcpy_inline` by introducing a `class CIR_MemCpyOp` with three separate operations. This approach is simple and clean.

If we were to use a single operation with different attributes, it would complicate the support for `__builtin_memcpy_inline` in the future. We would need to add an additional attribute (e.g., `inline`) to indicate the inline version of memcpy, and we would have to verify the validity of this attribute since `memmove` does not have an inline version. Moreover, this would make the lowering of `MemCpyOp` more complex, leading to nested if statements.

Reply:
@Lancern thanks for the clarifying my point. @PikachuHyA I agree with you on having separate names too.